### PR TITLE
Speed up `IsRegularPGroup`

### PR DESCRIPTION
**Pull request by fingolfin**
The speed up is achieved by carefully caching some more
intermediate results. Also avoid an Agemo call, and drop a size
check that I thought was beneficial, but in benchmarking harder
examples turned out to be not helpful.

Some timeing on my M1 MacBook Pro. Before this patch:

    gap> IsRegularPGroup(DirectProduct(SmallGroup(3^5,22),SmallGroup(3^5,22)));; time;
    27609
    gap> IsRegularPGroup(DirectProduct(SmallGroup(3^5,22),SmallGroup(3^5,39)));; time;
    77479

After this patch:

    gap> IsRegularPGroup(DirectProduct(SmallGroup(3^5,22),SmallGroup(3^5,22)));; time;
    1488
    gap> IsRegularPGroup(DirectProduct(SmallGroup(3^5,22),SmallGroup(3^5,39)));; time;
    3957

Also fix some references to Huppert's book.

### CODE REVIEW

**Comment by fingolfin** [comment]
@jackschmidt perhaps you'd be willing to give this patch a glance over?

File: lib/grp.gi [hunk 1]
```diff
@@ -433,43 +433,44 @@ InstallMethod( IsPowerfulPGroup,
...
+local p, hom, reps, as, bs, a, b, ap, bp, ab, ap_bp, ab_p, g, h, H;
 
   if not IsPGroup(G) then
     return false;
   fi;
 
   p:=PrimePGroup(G);
   if p = 2 then
-    # see [Hup67, Satz 10.3 a)]
+    # see [Hup67, Satz III.10.3 a)]
```
Comment (jackschmidt) [inline lib/grp.gi:444]: Good. The previous citation was incomplete. This is correct

File: lib/grp.gi [hunk 2]
```diff
@@ -483,27 +484,35 @@ local p, hom, reps, a, b, ap_bp, ab_p, H;
...
 
       # if the subgroup generated H by a and b is itself regular, we are also
-      # done. However we don't use recursion, here, as H may be equal to G;
-      # and also we have to be careful to not use too expensive code here.
-      # But a quick size check is certainly fine.
+      # done. However we don't use recursion here, as it is too expensive.
+      # we just check the direct definition, with a minor twist to avoid
+      # calling Agemo
+      g := ap_bp / ab_p;
+      h := Comm(a,b)^p;
```
Comment (jackschmidt) [inline lib/grp.gi:511]: Good. All the caching is correct and should avoid repeated element calculations

File: lib/grp.gi [hunk 3]
```diff
@@ -483,27 +484,35 @@ local p, hom, reps, a, b, ap_bp, ab_p, H;
...
+      if g = h or IsOne(g*h) then continue; fi;
```
Comment (jackschmidt) [inline lib/grp.gi:512]: Good. This new check is most advantageous for small p, but should always be cheap (g=h always cheap, g*h should not be expensive). For small p, I suspect it completely avoids the subgroup check in many cases

File: lib/grp.gi [hunk 4]
```diff
@@ -483,27 +484,35 @@ local p, hom, reps, a, b, ap_bp, ab_p, H;
...
       H := Subgroup(G, [a,b]);
-      if Size(H) <= p^p then continue; fi;
-
-      # finally the full check
-      H := DerivedSubgroup(H);
-      if not (ap_bp / ab_p) in Agemo(H, p) then
+      H := NormalClosure(H, [h]); # faster than Agemo(DerivedSubgroup(H), p);
+      if not g in H then
```
Comment (jackschmidt) [inline lib/grp.gi:515]: Good. This new H should be much faster to compute, and if g is not in H, then G is surely not regular, but if G is regular, then this H is the same as the more expensive Agemo subgroup, and the check will always pass.

**Review by jackschmidt — APPROVED** [review]
Thank you! These changes look good to me. I have not had a chance to time check them, or to try to find pathological examples, but the only line that potentially adds more work is the IsOne(g*h), and I hope that is negligible compared to everything else. All other changes should strictly reduce the amount of work done if the slow path is taken (and don't do anything if one of the earlier quick checks applies).

File: lib/grp.gi (hunk 3 above)
Comment (fingolfin) [reply lib/grp.gi:512]: Thank you very, very much for you feedback @jackschmidt and sorry for taking so long to get back to you.

Regarding this point, indeed `g=h` is fast and indeed for small $p$ the cost for checking `IsOne(g*h)` is absolutely worth it, based on extensive tests with groups of order $3^{10}$ formed as direct products of smaller $3$-groups.

File: lib/grp.gi (hunk 4 above)
Comment (fingolfin) [reply lib/grp.gi:515]: I've now added a comment explaining what's going on, I hope this makes it much clearer. Apologies for not doing that before, I really should have :-(

**Review by jackschmidt — APPROVED** [review]
Looks good, thanks! The revised changes continue to be correct and a strict improvement on the original, and the added documentation is much appreciated.

**Pull request merged**